proj-android/PowerAuthLibrary/src/main/java/io/getlime/security/powerauth/sdk/PowerAuthTokenStore.java

```diff
@@ -0,0 +1,421 @@
+                if (newTokenData != null) {
+                    // Store token data & report to listener
+                    storeTokenData(context, newTokenData);
+                    listener.onGetTokenSucceeded(new PowerAuthToken(PowerAuthTokenStore.this, newTokenData));
+                } else {
+                    // Report encryption error
+                    listener.onGetTokenFailed(new PowerAuthErrorException(PowerAuthErrorCodes.PA2ErrorCodeEncryptionError));
+                }
+            }
+
+            @Override
+            public void onNetworkError(Throwable t) {
+                // On failure, just notify listener about that error
+                listener.onGetTokenFailed(t);
+            }
+        });
+    }
+
+    /**
+     * Decrypt response received from server and create a new {@link PowerAuthPrivateTokenData} token data.
+     *
+     * @param decryptor A decryptor for response data decryption
+     * @param response An encrypted response
+     * @param tokenName Name of token
+     * @return Private token's data or null in case of failure.
+     */
+    private @Nullable PowerAuthPrivateTokenData decryptTokenData(ECIESEncryptor decryptor, TokenCreateResponse response, String tokenName) {
+        final ECIESCryptogram cryptogram = new ECIESCryptogram(response.getEncryptedData(), response.getMac());
+        final byte[] decryptedPayload = decryptor.decryptResponse(cryptogram);
+        if (decryptedPayload == null) {
+            return null;
+        }
+        TokenResponsePayload payload = httpClient.deserializePlainResponse(decryptedPayload, TokenResponsePayload.class);
+        if (payload.getTokenId() == null || payload.getTokenSecret() == null) {
+            return null;
+        }
+        final byte[] tokenSecretBytes = Base64.decode(payload.getTokenSecret(), Base64.NO_WRAP);
+        final PowerAuthPrivateTokenData tokenData = new PowerAuthPrivateTokenData(tokenName, payload.getTokenId(), tokenSecretBytes);
+        return tokenData.hasValidData() ? tokenData : null;
+    }
+
+
+    /**
+     * Remove previously created access token from the server and from local database.
+     * <p>
+     * Note that if the removal request doesn't succeed, then the local token's data is not removed.
+     * The method is thread safe, but it's not recommended to issue conflicting request for the same
+     * token's name in parallel (e.g. create &amp; remove token at the same time).
+     *
+     * @param context Context
+     * @param tokenName Name of token to be removed
+     * @param listener Listener with callbacks.
+     * @return {@code AsyncTask} associated with the running server request or null in case of error.
+     */
+    public @Nullable AsyncTask removeAccessToken(@NonNull final Context context, @NonNull final String tokenName, @NonNull final IRemoveTokenListener listener) {
+
+        Throwable error = null;
+        PowerAuthPrivateTokenData tokenData;
+
+        synchronized (this) {
```

Comment:
Note: Similar question as above - it seems we only try to synchronize "getTokenData" - can we handle it in the method itself?

Reply:
It would also work if I declare the method as synchronized, but it may be confusing for the API user, that the consistency is somehow guaranteed also during the internal network operation. That's not true and in fact, it's difficult to achieve (because we don't want to lock the whole object during the network task execution).

The whole `PowerAuthTokenStore` class is thread safe (e.g. you can't damage its internal data by accessing from multiple threads) but it's not aware agains the simultaneous asynchronous operations executed for the same token name. This is mentioned in the documentation.